DuckDuckGo/AnimatableTypingText.swift

```diff
@@ -0,0 +1,153 @@
+        _model = StateObject(wrappedValue: AnimatableTypingTextModel(text: text, onTypingFinished: onTypingFinished))
+        self.startAnimating = startAnimating
+        self.onTypingFinished = onTypingFinished
+    }
+
+    var body: some View {
+        ZStack(alignment: .topLeading) {
+            Text(text)
+                .frame(maxWidth: .infinity, alignment: .leading)
+                .visibility(.invisible)
+
+            if #available(iOS 15, *) {
+                Text(AttributedString(model.typedAttributedText))
+                    .frame(maxWidth: .infinity, alignment: .leading)
+            } else {
+                Text(model.typedAttributedText.string)
+                    .frame(maxWidth: .infinity, alignment: .leading)
+            }
+        }
+        .onChange(of: startAnimating.wrappedValue, perform: { shouldAnimate in
+            if shouldAnimate {
+                model.startAnimating()
+            } else {
+                model.stopAnimating()
+            }
+        })
+        .onAppear {
+            if startAnimating.wrappedValue {
+                model.startAnimating()
+            }
+        }
+    }
+}
+
+// MARK: - Model
+
+final class AnimatableTypingTextModel: ObservableObject {
+    private var timer: TimerInterface?
+
+    @Published private(set) var typedAttributedText: NSAttributedString = .init(string: "")
+
+    private var typingIndex = 0
+    private var textTypedSoFar: String = ""
+    private let text: String
+    private let onTypingFinished: (() -> Void)?
+    private let timerFactory: TimerCreating
+
+    init(text: String, onTypingFinished: (() -> Void)?, timerFactory: TimerCreating = TimerFactory()) {
+        self.text = text
+        self.onTypingFinished = onTypingFinished
+        self.timerFactory = timerFactory
+    }
+
+    func startAnimating() {
+        timer = timerFactory.makeTimer(withTimeInterval: 0.02, repeats: true, block: { [weak self] timer in
+            guard timer.isValid else { return }
+            self?.handleTimerEvent()
+        })
+    }
+
+    func stopAnimating() {
+        timer?.invalidate()
+        timer = nil
+
+        stopTyping()
+    }
+
+    deinit {
+        timer?.invalidate()
+        timer = nil
+    }
+
+    private func handleTimerEvent() {
+        if textTypedSoFar == text {
+            onTypingFinished?()
+            stopAnimating()
+            return
+        }
+
+        showCharacter()
+    }
+
+    private func stopTyping() {
+        typedAttributedText = NSAttributedString(string: text)
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
+            self?.onTypingFinished?()
+        }
+    }
+
+    private func showCharacter() {
+
+        func attributedTypedString(forTypedChars typedChars: [String.Element]) -> NSAttributedString {
+            guard #available(iOS 15, *) else {
+                return NSAttributedString(string: String(typedChars))
+            }
+
+            let chars = Array(text)
```

Comment:
This was existing logic. In macOS we use a different approach depending if we support macOS 12+ or not. If so we use an AttributedString and change the color of the characters we don’t show to clear. If macOS is lower than 12 w just we just display the character that is published each time.

Perhaps we can just use the same approach independently from the OS version? I wanted to double-check before removing the logic for the attributed string.

Also, since in macOS we will remove this code after the experiment, I haven’t moved this into BSK.

 wdyt?